[Screen Time Refactoring] Previously opened website won’t be blocked after relaunching Chrome

jesxilin opened the pull request:
#### 5bf1140d588362632333f571f622a2ed9862b9ea
<pre>
[Screen Time Refactoring] Previously opened website won’t be blocked after relaunching Chrome
<a href="https://bugs.webkit.org/show_bug.cgi?id=288148">https://bugs.webkit.org/show_bug.cgi?id=288148</a>
<a href="https://rdar.apple.com/145095487">rdar://145095487</a>

Reviewed by Aditya Keerthi.

Currently, we only install STWebpageController in `WebPageProxy::prepareToLoadWebPage`.
This is an issue for webpages or tabs that were already loaded.
When we&apos;ve navigated to a page, then quit Chrome and relaunch, the function
`WebPageProxy::prepareToLoadWebPage` does not get called again.
Chrome sets their session state and the webview&apos;s interaction state
which then calls into our WKWebView&apos;s `setInteractionState`that then calls
`WebPageProxy::restoreFromSessionState`.

Thus, we need to change where we call install STWebpageController. Instead, we will
create a controller if it does not exist when `didChangeScreenTimeWebpageControllerURL`
is called.

Also, we changed some API tests to wait until a controller is created as we do not
create a controller upon `WebPageProxy::prepareToLoadWebPage` but in
`WebPageProxy::didCommitLayerTree` which calls `didChangeScreenTimeWebpageControllerURL`.

Additionally, add an API test that mimics the state that Chrome is in.

* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::updateScreenTimeWebpageControllerURL):
(WebKit::PageClientImplCocoa::installScreenTimeWebpageController): Deleted.
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::installScreenTimeWebpageController): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::prepareToLoadWebPage):
(WebKit::WebPageProxy::hasVideoInPictureInPictureDidChange):
(WebKit::WebPageProxy::updatePlayingMediaDidChange):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(testShowsSystemScreenTimeBlockingView):
(TEST(ScreenTime, IsBlockedByScreenTimeTrue)):
(TEST(ScreenTime, IsBlockedByScreenTimeFalse)):
(TEST(ScreenTime, IsBlockedByScreenTimeMultiple)):
(TEST(ScreenTime, IsBlockedByScreenTimeKVO)):
(TEST(ScreenTime, IdentifierString)):
(TEST(ScreenTime, OffscreenSystemScreenTimeBlockingView)):
(TEST(ScreenTime, OffscreenBlurredScreenTimeBlockingView)):
(TEST(ScreenTime, DoNotDonateURLsInOffscreenWebView)):
(TEST(ScreenTime, ScreenTimeControllerInstalledAfterRestoreFromSessionState)):

Canonical link: <a href="https://commits.webkit.org/291022@main">https://commits.webkit.org/291022@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c670bb09ad859e8b810f95ed49e4ec91783d0d05

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91687 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11215 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/756 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96647 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42342 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93737 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11587 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19779 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70389 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27893 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94688 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8851 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83061 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50715 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8614 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/652 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41533 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78929 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/654 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98658 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18833 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79417 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19087 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78901 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78632 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23147 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/501 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11935 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14546 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18825 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24087 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18531 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21988 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20288 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3054,6 +3054,10 @@ void WebPageProxy::viewDidEnterWindow()
 #endif
+
+#if ENABLE(SCREEN_TIME)
+    m_pageClient->installScreenTimeWebpageController();
```
is m_pageClient guaranteed to not be null?

hortont424 replied:
protectedPageClient() is used unconditionally above (but also probably should be used here??)

pxlcoder replied:
Yes, this should use `protectedPageClient()`.

jesxilin replied:
I have another call in this file too that uses m_pageClient. Shall I change both to use `protectedPageClient()`?

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -291,7 +299,7 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
         @selector(setProfileIdentifier:),
         imp_implementationWithBlock(^(id object, NSString *profileIdentifier) {
-            identifier = profileIdentifier;
+            identifier = [profileIdentifier copy];
```
It would be better to simply make `identifier` a `RetainPtr`, since the `copy` here on its own is a memory leak.

pxlcoder reviewed (commented):
> Chrome calls `WebPageProxy::restoreFromSessionState`.

Commit message could be more specific about which API method Chrome is using, since it does not call `WebPageProxy` directly.

Also, can we add an API test that mimics what they are doing?

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
```diff
@@ -285,13 +293,13 @@ static BOOL systemScreenTimeBlockingViewIsPresent(TestWKWebView *webView)
         @selector(setProfileIdentifier:),
         imp_implementationWithBlock(^(id object, NSString *profileIdentifier) {
-            identifier = profileIdentifier;
+            identifier = [profileIdentifier copy];
```
To be clear, the copy should be removed now that you're using `RetainPtr`.
```suggestion
            identifier = profileIdentifier;
```

jesxilin replied:
Ah... I see 😓

jesxilin committed:
[Screen Time Refactoring] Previously opened website won’t be blocked after relaunching Chrome
https://bugs.webkit.org/show_bug.cgi?id=288148
rdar://145095487

Reviewed by Aditya Keerthi.

Currently, we only install STWebpageController in `WebPageProxy::prepareToLoadWebPage`.
This is an issue for webpages or tabs that were already loaded.
When we've navigated to a page, then quit Chrome and relaunch, the function
`WebPageProxy::prepareToLoadWebPage` does not get called again.
Chrome sets their session state and the webview's interaction state
which then calls into our WKWebView's `setInteractionState`that then calls
`WebPageProxy::restoreFromSessionState`.

Thus, we need to change where we call install STWebpageController. Instead, we will
create a controller if it does not exist when `didChangeScreenTimeWebpageControllerURL`
is called.

Also, we changed some API tests to wait until a controller is created as we do not
create a controller upon `WebPageProxy::prepareToLoadWebPage` but in
`WebPageProxy::didCommitLayerTree` which calls `didChangeScreenTimeWebpageControllerURL`.

Additionally, add an API test that mimics the state that Chrome is in.

* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::updateScreenTimeWebpageControllerURL):
(WebKit::PageClientImplCocoa::installScreenTimeWebpageController): Deleted.
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::installScreenTimeWebpageController): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::prepareToLoadWebPage):
(WebKit::WebPageProxy::hasVideoInPictureInPictureDidChange):
(WebKit::WebPageProxy::updatePlayingMediaDidChange):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(testShowsSystemScreenTimeBlockingView):
(TEST(ScreenTime, IsBlockedByScreenTimeTrue)):
(TEST(ScreenTime, IsBlockedByScreenTimeFalse)):
(TEST(ScreenTime, IsBlockedByScreenTimeMultiple)):
(TEST(ScreenTime, IsBlockedByScreenTimeKVO)):
(TEST(ScreenTime, IdentifierString)):
(TEST(ScreenTime, OffscreenSystemScreenTimeBlockingView)):
(TEST(ScreenTime, OffscreenBlurredScreenTimeBlockingView)):
(TEST(ScreenTime, DoNotDonateURLsInOffscreenWebView)):
(TEST(ScreenTime, ScreenTimeControllerInstalledAfterRestoreFromSessionState)):

Canonical link: https://commits.webkit.org/291022@main
